README
targz_sizes
===========

targz_sizes is a small and simple tool to find out how much space do specific
files take inside a .tar.gz file, after compression. It does that by reading and
decompressing the archive file in memory, then dumping raw information on the
standard output. Archive file is processed in small chunks, so that memory
usage is minimal.

This program has been written to check which directories take the most space in
my `duplicity` backups. This is different than checking space usage on a live
file system because of two factors: compression (text files often take very
little space in archives compared to live file system) and incremental backups
(files that change often are stored multiple times in backups). targz_sizes
performs only basic data gathering, you will need something to provide proper
statistics though.

Usage
-----

No options are implemented. The archive file is taken on standard input,
file sizes are printed on standard output. E.g.:

    targz_sizes <my_archive.tar.gz >file_sizes.txt

Output consists of lines, where each line shows size in bytes and name of file,
e.g.:

	145 targz_sizes-1.0/
	63 targz_sizes-1.0/AUTHORS
	3525 targz_sizes-1.0/missing
	42103 targz_sizes-1.0/configure
	93 targz_sizes-1.0/NEWS
	12936 targz_sizes-1.0/config.guess
	10025 targz_sizes-1.0/config.sub

Note that directory entries also take up space in the archive! Also note that
technically a file can be encoded by any number of bits, not necessarily being
round number of bytes. In such cases, amount of space shown in the output is
approximate.

Known problems
--------------

To compile you need to pass `--with-zlib` to ./configure. `zlib` is required,
even if the configure script goes well without this switch. E.g.:

    ./configure --with-zlib

Ustar-formatted file sizes are not supported, the output will be garbage. Ustar
format allows storing files bigger than 8GB in tar archives. GNU tar does not
use this format unless it is necessary, so if you used GNU tar and you know
there are no files bigger than 8GB, you're safe.

Standard input is not reopened in binary mode, as technically there is no way
to perform this operation with standard C primitives. This will be a problem on
platforms that by default provide standard input as a text stream (e.g.
Windows). In such cases, the output will be garbage.

The above problems should be easily fixable. You are free to fork the project
if you want to fix them.

Archive file names are not re-encoded to current locale. You can use `iconv`
for that.

Archive file names are truncated to 100 characters (the limit for original tar
specification). There are format extensions which can cope with longer names,
but those are not supported by targz_sizes. The output will contain the names
as they appear in the 100-byte field in the header, most probably simply
truncated.


NEWS
MORBO HAS NO NEWS FOR YOU!

src/targz_sizes.cpp
/*
 ============================================================================
 Name        : targz_sizes.cpp
 Author      : Tomasz Melcer
 Version     :
 Copyright   : GPL v3
 Description : Compute compressed sizes of individual files in a .tar.gz file
 ============================================================================
 */

#include <iostream>
#include <cstdio>
#include <zlib.h>

using namespace std;

#define COMPRESSED_BUFFER 102400

struct tarheader {
    char filename[100];
    char somejunk[24];
    char size[12];
    char otherjunk[512-100-24-12];
};

long long decode_octal(char* size) {
    long long value = 0;
    for (int i=0; i<11; i++) {
        value = value*8 + (size[i]&7);
    }
    return value;
}

int main(int argc, char** argv) {
    // argument parsing
    if (argc != 2) {
        cout << "Usage:" << endl
             << "   " << argv[0] << " file.tar.gz" << endl;

        return 1;
    }

    char* filename = argv[1];

    // opening files and setting up counters
    std::cout << "open " << filename << endl;
    FILE* file = fopen(filename, "rb");  // TODO: error checking

    unsigned char compressed[COMPRESSED_BUFFER];
    tarheader decompressed;
    gz_header gzip_header = {0};
    z_stream gzip_stream = {0};
    long long int bytes_read = 0;

#define ERROR " errmsg: " << (gzip_stream.msg ? gzip_stream.msg : "none") << endl

    gzip_stream.next_in = compressed;
    gzip_stream.avail_in = 0;
    gzip_stream.zalloc = Z_NULL;
    gzip_stream.zfree = Z_NULL;
    gzip_stream.next_out = (unsigned char*) &decompressed;
    gzip_stream.avail_out = sizeof(decompressed);
    std::cout << "inflateInit: " << inflateInit2(&gzip_stream, 15+32) << ERROR;  // TODO: error checking
    std::cout << "inflateGetHeader: " << inflateGetHeader(&gzip_stream, &gzip_header) << ERROR;
    //std::cout << "tar block size: " << sizeof(decompressed) << endl;

    // state machine
    int skip_blocks = 0;
    while(true) {
        //std::cout << "new cycle\n";
        // if stream's avail_in is empty, we need more data.
        if (gzip_stream.avail_in == 0) {
            //std::cout << "Empty input." << endl;
            // TODO: error checking
            // TODO: end of file condition
            size_t bytes = fread(&compressed, 1, COMPRESSED_BUFFER, file);
            //std::cout << "Read " << bytes << " bytes.\n";
            bytes_read += bytes;
            gzip_stream.next_in = compressed;
            gzip_stream.avail_in = bytes;
        }
        // post-condition: gzip_stream.avail_in > 0

        // if output buffer is not full, try to decompress more data
        if (gzip_stream.avail_out > 0) {
            //std::cout << "free space for decompression: " << gzip_stream.avail_out << endl;
            // TODO: error checking
            // TODO: end of stream condition
            int cond;
            std::cout << "inflate: " << (cond = inflate(&gzip_stream, Z_NO_FLUSH)) << ERROR;
            if (cond == 1)
                return 0;
        }

        // if output buffer is full, perform an action.
        if (gzip_stream.avail_out == 0) {
            //std::cout << "Full buffer. ";
            if (skip_blocks > 0) {
                decompressed.filename[20] = 0;
                //std::cout << "skipped: [[[" << (char*) &decompressed << "]]]" << endl;
                // we're inside a file. just ignore the block
                skip_blocks--;
            } else {
                // we've got a header. print details.
                // TODO: handle long file names.
                std::cout << "File: " << decompressed.filename
                          << ", decompressed size: " << decompressed.size
                          << ", position: " << gzip_stream.total_in
                          << endl;

                long long size = decode_octal(decompressed.size);
                skip_blocks = ((size + 511) / 512);

                //std::cout << "Blocks to skip: " << skip_blocks << " (from: " << size << ")" << endl;
            }

            gzip_stream.next_out = (unsigned char*) &decompressed;
            gzip_stream.avail_out = sizeof(decompressed);
        }
    }
}


src/targz_sizes.c
/*
 ============================================================================
 Name        : targz_sizes.c
 Author      : Tomasz Melcer
 Version     :
 License     : GPL v3
 Description : Compute compressed sizes of individual files in a .tar.gz file
 ============================================================================
 */

#include <stdint.h>
#include <stdio.h>

#include <zlib.h>

#define COMPRESSED_BUFFER 102400
#define TAR_BLOCK_SIZE 512

struct tarheader {
    char filename[100];
    char somejunk[24];
    char size[12];
    char otherjunk[512-100-24-12];
};

// technically 36 bits should be enough, but there is no uint_least36_t type
typedef uint_least64_t tarfilesize_t;

tarfilesize_t decode_octal(char* size) {
    tarfilesize_t value = 0;
    for (int i=0; i<11; i++) {
        value = value*8 + (size[i]&7);
    }
    return value;
}

int main(int argc, char** argv) {
    // argument parsing
    if (argc != 1) {
        fprintf(stderr, "Usage:\n   %s <file.tar.gz\n", argv[0]);
        return 1;
    }

    unsigned char compressed[COMPRESSED_BUFFER];
    unsigned char decompressed[TAR_BLOCK_SIZE];
    struct tarheader header;
    gz_header gzip_header = {0};
    z_stream gzip_stream = {0};
    long long int start_of_last_header = 0;

    gzip_stream.next_in = compressed;
    gzip_stream.avail_in = 0;
    gzip_stream.zalloc = Z_NULL;
    gzip_stream.zfree = Z_NULL;
    gzip_stream.next_out = (unsigned char*) &header;
    gzip_stream.avail_out = sizeof(header);
    inflateInit2(&gzip_stream, 15+32); // TODO: error checking
    inflateGetHeader(&gzip_stream, &gzip_header);

    // state machine
    int skip_blocks = 0;
    int return_code = 0;
    int more_data = 1;
    while(more_data) {
        // if stream's avail_in is empty, we need more data.
        if (gzip_stream.avail_in == 0) {
            size_t bytes = fread(&compressed, 1, COMPRESSED_BUFFER, stdin);
            if (bytes == 0 && ferror(stdin)) {
                perror("stdin");
                return_code = 1;
                break;
            } else if (bytes == 0 && feof(stdin)) {
                fprintf(stderr, "stdin: premature end of file\n");
                return_code = 1;
                break;
            }
            gzip_stream.next_in = compressed;
            gzip_stream.avail_in = bytes;
        }
        // post-condition: gzip_stream.avail_in > 0

        // if output buffer is not full, try to decompress more data
        if (gzip_stream.avail_out > 0) {
            int cond = inflate(&gzip_stream, Z_NO_FLUSH);
            if (cond == Z_STREAM_END) {
                // this is the last iteration of the loop
                more_data = 0;
            } else if (cond < 0) {
                fprintf(stderr, "stdin: %s\n", gzip_stream.msg);
                return_code = 1;
                break;
            }
        }

        // if output buffer is full, perform an action.
        if (gzip_stream.avail_out == 0) {
            if (skip_blocks > 0) {
                // we're inside a file. just ignore the block
                skip_blocks--;
            } else {
                // we've got a header
                tarfilesize_t size = decode_octal(header.size);
                skip_blocks = ((size + 511) / 512);
            }

            if (skip_blocks > 0) {
                gzip_stream.next_out = (unsigned char*) &decompressed;
                gzip_stream.avail_out = sizeof(decompressed);
            } else {
                // tar uses blocks with null filename to mark end of stream.
                // ignore them.
                // TODO: handle long file names.
                if (header.filename[0] != 0) {
                    // to ensure the filename is 0-terminated somewhere (note:
                    // we're not handling longer filenames than those which
                    // fit the original tar limits).
                    header.somejunk[0] = 0;
                    printf("%lld %s\n",
                        gzip_stream.total_in - start_of_last_header,
                        header.filename);
                }

                start_of_last_header = gzip_stream.total_in;

                gzip_stream.next_out = (unsigned char*) &header;
                gzip_stream.avail_out = sizeof(header);
            }
        }
    }

    // free resources and make valgrind happy
    inflateEnd(&gzip_stream);

    return return_code;
}
